File: Easy/771 - JewelsAndStones/Solution.cs
using System.Collections.Generic;
/// Question 771. Jewels and Stones (https://leetcode-cn.com/problems/jewels-and-stones/)
/// Difficulty: Easy

public class Solution
{
    public int NumJewelsInStones(string J, string S)
    {
        return NumJewelsInStones(J, S);
    }

    public int NumJewelsInStonesIndexOf(string J, string S)
    {
        // 执行用时: 96 ms, 在所有 C# 提交中击败了 93.98% 的用户
        // 内存消耗: 23.2 MB, 在所有 C# 提交中击败了 5.00% 的用户
        int jewelsCount = 0;
        foreach (char s in S)
        {
            if (J.IndexOf(s) >= 0)
            {
                jewelsCount++;
            }
        }
        return jewelsCount;
    }

    public int NumJewelsInStonesLinq(string J, string S)
    {
        // 执行用时: 80 ms, 在所有 C# 提交中击败了 93.98% 的用户
        // 内存消耗: 23.2 MB, 在所有 C# 提交中击败了 5.00% 的用户
        int jewelsCount = 0;
        foreach (char s in S)
        {
            if (J.Contains(s))
            {
                jewelsCount++;
            }
        }
        return jewelsCount;
    }

    public int NumJewelsInStonesSet(string J, string S)
    {
        // 执行用时：92 ms, 在所有 C# 提交中击败了 45.61% 的用户
        // 内存消耗：22.7 MB , 在所有 C# 提交中击败了 65.69% 的用户
        HashSet<char> jewels = new HashSet<char>();
        foreach (char jewel in J)
        {
            jewels.Add(jewel);
        }

        int jewelsCount = 0;
        foreach (char stone in S)
        {
            if (jewels.Contains(stone))
            {
                jewelsCount++;
            }
        }

        return jewelsCount;
    }
}

File: Easy/771 - JewelsAndStones/solution.ts
/**
 * Question 771. Jewels and Stones (https://leetcode-cn.com/problems/jewels-and-stones/)
 * 执行用时：96 ms, 在所有 typescript 提交中击败了 22.86% 的用户
 * 内存消耗：39.3 MB, 在所有 typescript 提交中击败了 12.50% 的用户
 *
 * @param {string} J
 * @param {string} S
 * @returns {number}
 */
function numJewelsInStones(J: string, S: string): number {
  const jewels: Map<string, number> = new Map<string, number>();
  for (let jewel of J) {
    jewels.set(jewel, 1);
  }

  let jewelCount: number = 0;

  for (let stone of S) {
    if (jewels.has(stone)) {
      jewelCount++;
    }
  }

  return jewelCount;
}

/**
 * 执行用时：96 ms, 在所有 typescript 提交中击败了 22.86% 的用户
 * 内存消耗：39.5 MB, 在所有 typescript 提交中击败了 12.50% 的用户
 *
 * @param {string} J
 * @param {string} S
 * @returns {number}
 */
function numJewelsInStonesUsingSet(J: string, S: string): number {
  const jewels: Set<string> = new Set<string>();
  for (let jewel of J) {
    jewels.add(jewel);
  }

  let jewelCount: number = 0;

  for (let stone of S) {
    if (jewels.has(stone)) {
      jewelCount++;
    }
  }

  return jewelCount;
}
